build.c

```diff
@@ -541,24 +551,71 @@ queryload(void)
 #endif
 }
 
+/* returns remaining GNU/tokens */
+static void
+returngmake(void)
+{
+	if (buildopts.gmakepipe[1] >= 0) {
+		if (write(buildopts.gmakepipe[1], gmaketokens, gmakelatest - gmaketokens) < 0) {
+			warn("last write to jobserver:");
```

Comment:
this is called in a signal handler (termsignal, below), but warn() then uses the printf(3) family, which is not signal-safe. isn't this technically illegal? (it exits immediately after, so maybe that's ok?)

Reply:
I've moved the warn() part to a separate function.